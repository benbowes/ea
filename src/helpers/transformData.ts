import { Input, Output } from "../types";

const sortCaseInsensitive = <T extends string>(array: T[]): T[] => {
  return [...array].sort((a, b) => {
    return a.toString().toLowerCase().localeCompare(b.toString().toLowerCase());
  });
};

const sortBands = (array: Record<"name", string>[]) => {
  return [...array].sort((a, b) => {
    return a["name"]
      .toString()
      .toLowerCase()
      .localeCompare(b["name"].toString().toLowerCase());
  });
};

const uniqueFlatArray = <T>(items: T[]): T[] => [
  ...Array.from(new Set(items.map((item) => item))),
];

export function transformData(input: Input[]): Output[] {
  const getRecordLabels = (inputData: Input[]) => {
    const recordLabels = inputData.reduce((acc: string[], recordLabel) => {
      return [
        ...acc,
        ...(recordLabel?.bands || []).reduce(
          (acc: string[], band) => [...acc, band?.recordLabel ?? ""],
          []
        ),
      ];
    }, []);

    const uniqueSortedRecordLabels = sortCaseInsensitive<string>(
      uniqueFlatArray<string>(recordLabels)
    );

    return uniqueSortedRecordLabels.map((recordLabel) => recordLabel);
  };

  const recordLabels = getRecordLabels(input);

  return recordLabels.reduce((acc: any, recordLabel) => {
    const bandList = input.reduce((acc: any, responseItem: any) => {
      const found = responseItem.bands.find(
        (band: any) => band.recordLabel === recordLabel
      )?.name;

      const festivals = input.reduce((acc: any, inputItem) => {
        const festival = (inputItem.bands || []).find(
          (band) => band.name === found
        );
        return festival && inputItem.name ? [...acc, inputItem.name] : acc;
      }, []);

      const festivalsSorted = sortCaseInsensitive<string>(festivals);

      return found
        ? [...acc, { name: found, festivals: festivalsSorted }]
        : acc;
    }, []);

    const sortedBandList = sortBands(bandList);

    return [...acc, { recordLabel, bands: sortedBandList }];
  }, []) as Output[]; // Lazy typing
}
