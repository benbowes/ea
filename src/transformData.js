const sortCaseInsensitive = (array, field) => {
  return array.sort((a, b) => {
    const A = !field ? a : a[field];
    const B = !field ? b : b[field];
    return A.toLowerCase().localeCompare(B.toLowerCase());
  });
};

const uniqueFlatArray = (items) => [
  ...Array.from(new Set(items.map((item) => item))),
];

export function getOutput(input) {
  const getSortedRecordLabels = (inputData) => {
    const recordLabels = inputData.reduce((acc, recordLabel) => {
      return [
        ...acc,
        ...recordLabel.bands.reduce(
          (acc, band) => [...acc, band.recordLabel],
          []
        ),
      ];
    }, []);

    const uniqueSortedRecordLabels = sortCaseInsensitive(
      uniqueFlatArray(recordLabels)
    );

    return uniqueSortedRecordLabels.map((recordLabel) => recordLabel);
  };

  return getSortedRecordLabels(input).reduce((acc, recordLabel) => {
    const bandList = input.reduce((acc, responseItem) => {
      const found = responseItem.bands.find(
        (band) => band.recordLabel === recordLabel
      )?.name;

      const festivals = input.reduce((acc, inputItem) => {
        const festival = inputItem.bands.find((band) => band.name === found);
        return festival && inputItem.name ? [...acc, inputItem.name] : acc;
      }, []);

      const festivalsSorted = sortCaseInsensitive(festivals);

      return found
        ? [...acc, { name: found, festivals: festivalsSorted }]
        : acc;
    }, []);

    const sortedBandList = sortCaseInsensitive(bandList, "name");

    return [...acc, { recordLabel, bands: sortedBandList }];
  }, []);
}
