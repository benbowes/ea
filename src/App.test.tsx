import { input } from "./input";
import { getOutput } from "./transformData";

test("Given expected input, output is as expected", () => {
  expect(getOutput(input)).toMatchSnapshot();
});
