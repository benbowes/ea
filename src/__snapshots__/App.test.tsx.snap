// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given expected input, output is as expected 1`] = `
Array [
  Object {
    "bands": Array [
      Object {
        "festivals": Array [
          "LOL-palooza",
        ],
        "name": "Winter Primates",
      },
    ],
    "recordLabel": "",
  },
  Object {
    "bands": Array [
      Object {
        "festivals": Array [],
        "name": "Critter Girls",
      },
      Object {
        "festivals": Array [
          "Trainerella",
        ],
        "name": "Manish Ditch",
      },
    ],
    "recordLabel": "ACR",
  },
  Object {
    "bands": Array [
      Object {
        "festivals": Array [
          "Trainerella",
        ],
        "name": "YOUKRANE",
      },
    ],
    "recordLabel": "Anti Records",
  },
  Object {
    "bands": Array [
      Object {
        "festivals": Array [
          "LOL-palooza",
        ],
        "name": "Jill Black",
      },
      Object {
        "festivals": Array [
          "Small Night In",
        ],
        "name": "The Black Dashes",
      },
    ],
    "recordLabel": "Fourth Woman Records",
  },
  Object {
    "bands": Array [
      Object {
        "festivals": Array [
          "Small Night In",
        ],
        "name": "Green Mild Cold Capsicum",
      },
    ],
    "recordLabel": "Marner Sis. Recording",
  },
  Object {
    "bands": Array [
      Object {
        "festivals": Array [
          "Small Night In",
        ],
        "name": "Yanke East",
      },
    ],
    "recordLabel": "MEDIOCRE Music",
  },
  Object {
    "bands": Array [
      Object {
        "festivals": Array [
          "Trainerella",
        ],
        "name": "Adrian Venti",
      },
    ],
    "recordLabel": "Monocracy Records",
  },
  Object {
    "bands": Array [
      Object {
        "festivals": Array [
          "Small Night In",
        ],
        "name": "Squint-281",
      },
    ],
    "recordLabel": "Outerscope",
  },
  Object {
    "bands": Array [
      Object {
        "festivals": Array [
          "LOL-palooza",
        ],
        "name": "Frank Jupiter",
      },
      Object {
        "festivals": Array [],
        "name": "Propeller",
      },
    ],
    "recordLabel": "Pacific Records",
  },
  Object {
    "bands": Array [
      Object {
        "festivals": Array [
          "Small Night In",
          "Trainerella",
        ],
        "name": "Wild Antelope",
      },
    ],
    "recordLabel": "Still Bottom Records",
  },
  Object {
    "bands": Array [
      Object {
        "festivals": Array [
          "LOL-palooza",
        ],
        "name": "Werewolf Weekday",
      },
    ],
    "recordLabel": "XS Recordings",
  },
]
`;
